INGInious: provide a NixOS module

layus opened the pull request:
###### Motivation for this change
- Add support for INGInious in NixOS
- Cleanup & update INGInious pacakge
###### Things done
- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

You may want to try it, it is quite a nice soft :-).

layus committed:
python-lti: 0.4.0 -> 0.4.1

layus committed:
inginious: turn lib into a python app.

layus committed:
inginious: init NixOS module

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+in
+{
+  options.services.lighttpd.inginious = {
+    enable = mkOption {
```
You could use `mkEnableOption` here

bjornfor replied:
... but where would the useful doc text "[...] an automated code testing and grading system" be located then?

groxxda replied:
`mkEnableOption "INGInious, an automated code testing and grading system"`
[expands `description` to](https://github.com/NixOS/nixpkgs/blob/master/lib/options.nix#L30)
`"Whether to enable INGInious, an automated code testing and grading system."`

bjornfor replied:
Ah, I didn't know! Thanks.

layus replied:
Yep, I can do that :-).

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+
+    tasksDirectory = mkOption {
+      type = types.path;
+      default = "${pkgs.inginious}/lib/python2.7/site-packages/inginious/tasks";
```
Does Inginious treat these files as readonly? Otherwise it might be good to copy the tasks to something like `/var/lib/INGInious/tasks`

layus replied:
It makes little sense to have a readonly tasks dir, true, but it is enough for demo purposes.
It is howerver a valid default value that will work in any case without addin extra dependencies and allows to test the install before going further.

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+    remoteAgents = mkOption {
+      type = types.listOf (types.attrsOf types.str);
+      default = [];
+      example = [ { host = "145.33.17.89"; port = "1345"; } ];
```
Please use one of the [recommended ](https://tools.ietf.org/html/rfc5737) documentation prefixes:
```
3.  Documentation Address Blocks

   The blocks 192.0.2.0/24 (TEST-NET-1), 198.51.100.0/24 (TEST-NET-2),
   and 203.0.113.0/24 (TEST-NET-3) are provided for use in
   documentation.
```

layus replied:
Not sure what you mean here...

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+          $HTTP["host"] =~ "${cfg.hostPattern}" {
+            fastcgi.server = ( "/${execName}" =>
+              ((
+                "socket" => "/tmp/fastcgi.socket",
```
Is this an upstream recommendation? Otherwise I'd prefer the socket to live in `/run/`

layus replied:
This is the example given by web.py and lighttpd. I see no reason why it could not be moved, but I am not sure that changing the value here is enough. I need investigations.

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+                "bin-path" => "${inginious}/bin/${execName} --config=${inginiousConfigFile}",
+                "max-procs" => 1,
+                "bin-environment" => (
+                  ${lib.optionalString false ''"DOCKER_HOST" => "tcp://192.168.59.103:2375",'' }
```
Is this intentional? Looks to me as if the condition is missing or the code is superfluous :)

layus replied:
Yep, missed leftover from an old attempt :-)

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+            url.rewrite-once = (
+              "^/.well-known/.*" => "$0",
+              "^/static/.*" => "$0",
+              #"^${inginious}/" => "$0",
```
Also superfluous?

layus replied:
If I remember well, I need this one to avoid rewriting aliases defined below.

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+          }
+        '';
+
+        systemd.services.lighttpd.preStart = ''
```
I'm pretty sure `preStart` is not merged. Might be better to create a new systemd service that takes care of doing this with correct ordering (`after` or `before`).
I also think that without explicit ordering the `lighttpd.preStart` will be executed in parallel to the startup of the wanted `docker.service`. Since I don't know much about neither docker nor inginious this may or may not be a problem :wink:

layus replied:
You sir are completely right :smile:. I will fix this.

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+        '';
+        systemd.services.lighttpd.serviceConfig.TimeoutStartSec = "30m";
+
+        systemd.services.lighttpd.wants = [ "docker" ];
```
Please use `docker.service` here or whatever fits

layus replied:
Indeed.

groxxda commented on pkgs/servers/inginious/default.nix:
```diff
@@ -0,0 +1,76 @@
+  version = "0.3a2.dev0";
+  name = "inginious-${version}";
+
+  disabled = pythonPackages.isPy3k;
```
Since this is tied to `python27Packages` wouldn't it be more appropriate to set `disabled = ! pythonPackages.isPy27`?

layus commented:
The build will fail as PyLTI is not working with pytest_29, but it is "not my fault" :wink: Is there some way to fix it ? Anway, I made all the requested changes.

@groxxda Would you mind looking at 6440241 again ?

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+      '';
+    };
+
+    local = mkOption {
```
Since you mention that the `backend`-type `remote` might be implemented in the future, it would be preferable to keep the API upward-compatible here:
Please rename `local` to something like `backendType`, change it's type to `types.enum [ 'local' 'remote_manual']`.
Don't forget to adjust at the top of this file and also `remoteAgents.description`

groxxda commented on nixos/modules/services/web-servers/lighttpd/inginious.nix:
```diff
@@ -0,0 +1,257 @@
+        systemd.services.lighttpd.after = [ "mongodb.service" "docker.service" ];
+
+        systemd.services.inginious-prefetch = {
+          script = let
```
The indentation looks bogous to me :wink:

layus commented:
@groxxda, I renamed local to backendType as you suggested. I could not do much about the indentation however, as it seems right to me.
I also moved the docker-sync service to the section specific to the local backend.

layus commented:
Found a way to fix PyLTI, now travis should run fine :-)

layus commented:
Travis checks passed. I think this is ready to get merged, unless someone has more comments?

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6952,19 +6952,22 @@ in modules // {
-  lti = buildPythonPackage rec {
-    version = "0.4.0";
+  lti = let self' = (self.override {self = self';}) // {pytest = self.pytest_27;};
+  in buildPythonPackage rec {
```
If some package needs pytest 2.7 instead of the current 2.9, then fix that package and set it to explicitly pytest_27.
pytest is typically just a buildInput and not a propagatedBuildInput so there won't be any problems with multiple versions of the same package.

FRidh commented:
apologies for responding this late. As I wrote just now, you can use pytest_27 instead of pytest for PyLTI if that fixes the package.

layus commented on pkgs/top-level/python-packages.nix:
```diff
@@ -6952,19 +6952,22 @@ in modules // {
+    buildInputs = with self'; [
       flask httpretty oauthlib pyflakes pytest pytestcache pytestcov covCore
       pytestflakes pytestpep8 sphinx mock
     ];
```
Reply to https://github.com/NixOS/nixpkgs/pull/16086/files#r67858876:
It does not work because the buildInputs also contain `pytestcache pytestcov pytestflakes pytestpep8`. As these depend on pytest_29, a conflict occurs if the override is not deep enough.
That being said, I override pytest only for buildInputs, not propagatedBuildInputs, and while such an override looks big, it is really a very small overhead thanks to lazyness.

layus commented:
@FRidh, I think you missed the real difficulty here. See my description above.

FRidh commented:
@layus yep, it's clear to me now. It's a pity it has to be like this but so be it. 
Since nobody else had any comments I'll merge it.
